dotnet: use unpacked packages in store

corngood opened the pull request:
## Description of changes

This is still a bit of a WIP, although it does pass nixpkgs-review on linux and darwin, and fetch-deps works on everything in nixpkgs.  I mostly just want to do some more cleanup passes and pull out some incidental changes into separate commits.

The idea here is that we keep packages in `$out/share/nuget/packages/[name]/[version]`, unpacked, in a structure that's compatible with what you'd have in `~/.nuget/packages`.  Then in a hook, we point `$NUGET_FALLBACK_PACKAGES` at all the dependencies.  We also create a source, which contains 'stub' `.nupkg` files, containing only the `.nuspec`, and add it to the root `nuget.config`.

There are still some rough edges.  For example, `dotnet tool restore` doesn't really play nice with `$NUGET_PACKAGES`, and we still need to treat the package directory as a 'source' for projects that don't have locked dependencies.  In order to deal with this, we pass `installable = true` to `mkNugetDeps`, which tells it to create full `.nupkg` files instead of stubs.

Other highlights:

- we patch (e.g. interpreter, rpaths, shebangs) all packages in nix-store, rather than after restore
- fetch-deps is now more closely based on the package derivation
- we strip existing nuget.config files (except in fetch-deps), and create a new one in the build root
- output packages are always unpacked (even if a .nupkg also exists), which stops runtime dependencies from being obfuscated
- `projectReferences` is now obsolete, and packages can be used from any build input

FYI @NixOS/dotnet @Smaug123 (aside: would you like to be added to the dotnet team?)

Questions:

- should we leave `.nupkg` generation for (dependent) build time, and keep them out of the store altogether?

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

corngood committed:
dotnet: use unpacked nuget packages

corngood committed:
buildDotnetModule: remove nugetDeps from derivation args

This allows fetch-deps to work even when the deps file is missing.

corngood committed:
nuget-to-nix: catch errors from process substitution

corngood committed:
nexusmods-app: fix intermittent test failure

corngood commented:
Result of `nixpkgs-review pr 327651` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>141 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>alttpr-opentracker</li>
    <li>am2rlauncher</li>
    <li>avalonia-ilspy</li>
    <li>azure-functions-core-tools</li>
    <li>beatsabermodmanager</li>
    <li>bicep</li>
    <li>bililiverecorder</li>
    <li>blendfarm</li>
    <li>boogie</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>cavalier</li>
    <li>celeste64</li>
    <li>certdump</li>
    <li>csharp-ls</li>
    <li>csharpier</li>
    <li>csharprepl</li>
    <li>cyclonedx-cli</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>dotnet-aspnetcore</li>
    <li>dotnet-aspnetcore_7</li>
    <li>dotnet-aspnetcore_8</li>
    <li>dotnet-outdated</li>
    <li>dotnet-runtime</li>
    <li>dotnet-runtime_7</li>
    <li>dotnet-runtime_8</li>
    <li>dotnet-sdk</li>
    <li>dotnet-sdk_7</li>
    <li>dotnet-sdk_8</li>
    <li>dotnetCorePackages.aspnetcore_9_0</li>
    <li>dotnetCorePackages.dotnet_8.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_8.runtime</li>
    <li>dotnetCorePackages.dotnet_8.sdk</li>
    <li>dotnetCorePackages.dotnet_8.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_8.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_8.vmr</li>
    <li>dotnetCorePackages.dotnet_9.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_9.runtime</li>
    <li>dotnetCorePackages.dotnet_9.sdk</li>
    <li>dotnetCorePackages.dotnet_9.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_9.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_9.vmr</li>
    <li>dotnetCorePackages.patch-nupkgs</li>
    <li>dotnetCorePackages.runtime_9_0</li>
    <li>dotnetCorePackages.sdk_6_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_3xx</li>
    <li>dotnetCorePackages.sdk_8_0_1xx</li>
    <li>dotnetCorePackages.sdk_9_0</li>
    <li>eventstore</li>
    <li>fable</li>
    <li>famistudio</li>
    <li>fantomas</li>
    <li>formula</li>
    <li>fsautocomplete</li>
    <li>galaxy-buds-client</li>
    <li>garnet</li>
    <li>gh-gei</li>
    <li>git-credential-manager</li>
    <li>github-runner</li>
    <li>gitversion</li>
    <li>godot3-mono</li>
    <li>godot3-mono-debug-server</li>
    <li>godot3-mono-export-templates</li>
    <li>godot3-mono-headless</li>
    <li>godot3-mono-headless.dev</li>
    <li>godot3-mono-headless.man</li>
    <li>godot3-mono-server</li>
    <li>godot3-mono.dev</li>
    <li>godot3-mono.man</li>
    <li>ilspycmd</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>jellyfin</li>
    <li>jetbrains.clion</li>
    <li>jetbrains.rider</li>
    <li>kavita</li>
    <li>knossosnet</li>
    <li>kryptor</li>
    <li>libation</li>
    <li>lidarr</li>
    <li>lubelogger</li>
    <li>lumafly</li>
    <li>marksman</li>
    <li>mqttmultimeter</li>
    <li>msbuild</li>
    <li>naps2</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>networkminer</li>
    <li>nexusmods-app</li>
    <li>nexusmods-app-unfree</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>openra</li>
    <li>opentabletdriver</li>
    <li>openutau</li>
    <li>osu-lazer</li>
    <li>pablodraw</li>
    <li>parabolic</li>
    <li>pbm</li>
    <li>pinta</li>
    <li>pre-commit</li>
    <li>pre-commit.dist</li>
    <li>prowlarr</li>
    <li>ps3-disc-dumper</li>
    <li>pupdate</li>
    <li>python311Packages.asteroid-filterbanks</li>
    <li>python311Packages.asteroid-filterbanks.dist</li>
    <li>python311Packages.clr-loader</li>
    <li>python311Packages.clr-loader.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pythonnet</li>
    <li>python311Packages.pythonnet.dist</li>
    <li>python312Packages.asteroid-filterbanks</li>
    <li>python312Packages.asteroid-filterbanks.dist</li>
    <li>python312Packages.clr-loader</li>
    <li>python312Packages.clr-loader.dist</li>
    <li>python312Packages.pythonnet</li>
    <li>python312Packages.pythonnet.dist</li>
    <li>radarr</li>
    <li>readarr</li>
    <li>retrospy</li>
    <li>roslyn</li>
    <li>roslyn-ls</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>slskd</li>
    <li>sonarr</li>
    <li>space-station-14-launcher</li>
    <li>technitium-dns-server</li>
    <li>tone</li>
    <li>torrentstream</li>
    <li>vintagestory</li>
    <li>wasabibackend</li>
    <li>xivlauncher</li>
  </ul>
</details>

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-check-hook.sh:
```diff
@@ -42,24 +42,26 @@ dotnetCheckHook() {
-        if [[ $projectFile == *.csproj ]]; then
-            runtimeIdFlagsArray=("--runtime" "$dotnetRuntimeId")
-        fi
+        for runtimeId in "${dotnetRuntimeIdsArray[@]}"; do
```
it was for completeness, but in theory you could run multiple cpu archs on the same OS

GGG-KILLER replied:
I see, hadn't thought of that, my bad!

corngood commented on pkgs/development/compilers/dotnet/install-nupkgs-hook.nix:
I should have left that out for now, but I'd like to generalise the handling of nupkg files.  Something like:

installPhase -> `dotnet pack`
preFixup -> unpack
postFixup -> pack

so then anything done to the packages in `fixupPhase` would be reflected in the unpacked output package and also the corresponding .nupkg, if that's required.

GGG-KILLER replied:
I see, makes sense, would allow people to avoid this boilerplate

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/fetch-deps.sh:
```diff
@@ -1,111 +1,22 @@
-
-runtimeIds=(@runtimeIds@)
-
-for rid in "${runtimeIds[@]}"; do
```
That's here now:

https://github.com/NixOS/nixpkgs/pull/327651/files#diff-85d3cfd845d109ea11a8b1b45fc62bedd5d6f320185b8608ca1da8362c92b69fR193

GGG-KILLER replied:
Ah fair enough, didn't make the connection between those two

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/fetch-deps.sh:
```diff
@@ -1,111 +1,22 @@
 tmpFile="$tmp"/deps.nix
 echo -e "# This file was automatically generated by passthru.fetch-deps.\n# Please dont edit it manually, your changes might get overwritten!\n" > "$tmpFile"
-nuget-to-nix "$tmp/nuget_pkgs" "$tmp/excluded_list" >> "$tmpFile"
+@nugetToNix@/bin/nuget-to-nix "$NUGET_PACKAGES" >> "$tmpFile"
```
That's now in the sdk setup hook: 

https://github.com/NixOS/nixpkgs/pull/327651/files#diff-e6f6864e4e42982575fd661a7af06d76d9f22ab876397962c1875e2f4984a7c8R11

GGG-KILLER replied:
Ah ok, didn't know how setup hooks worked, my bad

corngood commented on pkgs/build-support/dotnet/build-dotnet-module/fetch-deps.sh:
```diff
@@ -1,111 +1,22 @@
-
-export DOTNET_NOLOGO=1
-export DOTNET_CLI_TELEMETRY_OPTOUT=1
+phases="
+    ${prePhases[*]:-}
+    unpackPhase
+    patchPhase
+    ${preConfigurePhases[*]:-}
+    configurePhase
+" genericBuild
```
Yeah, it's based on the phase documentation in 

> ### Building a `stdenv` package in `nix-shell` {#sec-building-stdenv-package-in-nix-shell}

I did something similar for fetching deps in VMR.

Alternatively we could override some phases and run all of `genericBuild`.  I slightly prefer this way so far.

GGG-KILLER replied:
Oh I see, my knowledge in nix was lacking on this, pretty interesting approach to avoid code duplication

corngood committed:
buildDotnetModule: fix typo

GGG-KILLER commented:
Looks good to me so far.

> we currently output .nupkg files next to the unpacked packages (as in ~/.nuget/packages). could this be avoided?

Maybe if in `fetch-deps` we generated a nuget lockfile we could avoid this, but other than that I don't see much of an alternative.

> maybe the .nupkgs can be empty and still function as a source for the purpose of version detection?

Maybe if we keep only the nuspec file in there we could somehow hack the restore algo into supporting it, but I don't know how I feel about it, feels a bit flaky.

> maybe source and packages should be separate outputs?

I like this idea, I think it'd make things less confusing as one is supposed to be the target directory and the other a source.

corngood commented:
> Maybe if we keep only the nuspec file in there we could somehow hack the restore algo into supporting it, but I don't know how I feel about it, feels a bit flaky.

I'll do some more testing around this.  What's in here right now might not be required for all packages.  We could possibly allow multiple options (best first):

1. add all dependency package dirs to `fallbackPackageFolders` in nuget.config
2. create `NUGET_FALLBACK_PACKAGES` with symlinks to dependencies
3. create `NUGET_PACKAGES` with symlinks to dependencies
4. do (3) and also add a package source with all the dependencies .nupkgs (pretty much what this PR currently does)

> maybe source and packages should be separate outputs?

One problem with this is that it obfuscates runtime deps, unless we make the .nupkgs uncompressed, or extract the dependencies.

GGG-KILLER commented:
> I'll do some more testing around this. What's in here right now might not be required for all packages. We could possibly allow multiple options (best first):
> 
> 1. add all dependency package dirs to `fallbackPackageFolders` in nuget.config
> 2. create `NUGET_FALLBACK_PACKAGES` with symlinks to dependencies
> 3. create `NUGET_PACKAGES` with symlinks to dependencies
> 4. do (3) and also add a package source with all the dependencies .nupkgs (pretty much what this PR currently does)

I might be mistaken, but 1 and 2 seem to be the same? They do seem like the best option considering that it's immutable and `dotnet restore` won't try to modify them, but considering we don't want any networking at all in `dotnet restore` and already have all the dependencies listed and downloaded, I don't see much of a difference between 1, 2 and 3 as in theory none of them would modify the source dir.

Given that most projects don't have lockfiles at all, I don't think we'll be able to get rid of overriding the source with a directory full of nupkgs as it'll try to use online sources instead if we don't do it.

For now I think the most realistic path would be to proceed with 4 as it works on projects without lockfiles and maybe look into requiring lockfiles for projects or something similar so that we can get rid of the directory full of `.nupkg`s entirely.

corngood commented:
> I might be mistaken, but 1 and 2 seem to be the same in practice?

Yeah, more or less.  It's just that `NUGET_FALLBACK_PACKAGES` only supports a single dir.

> They do seem like the best option considering that it's immutable and dotnet restore won't try to modify them,

Currently `nuget-to-nix` has to search through one package dir with a mix of built and downloaded packages, so in theory it would simplify that.

> Given that most projects don't have lockfiles at all, I don't think we'll be able to get rid of overriding the source with a directory full of nupkgs as it'll try to use online sources instead if we don't do it.

One thing I'm not sure of, and was going to test, is if we can do something like put empty files in place of the .nupkg files.  If it's only using the source to determine which versions are available, maybe that'll work?

Another crazier option might be to make a fake http source using a daemon?

GGG-KILLER commented:
> Yeah, more or less. It's just that `NUGET_FALLBACK_PACKAGES` only supports a single dir.

It actually supports multiple, but they need to be separated by semicolons:

> The value must contain the full path or paths to the fallback package folders. Multiple paths may be included by setting the value to a `;` delimited list.


> Currently `nuget-to-nix` has to search through one package dir with a mix of built and downloaded packages, so in theory it would simplify that.

How so? I thought it only searched through stuff that `dotnet restore` pulled, so if there's any dependencies on csprojs or nix-built stuff then there wouldn't be any `.nupkg`s there for it to find.

> One thing I'm not sure of, and was going to test, is if we can do something like put empty files in place of the .nupkg files. If it's only using the source to determine which versions are available, maybe that'll work?
> 
> Another crazier option might be to make a fake http source using a daemon?

I've asked a question in the .NET Discord to the nuget folks about it, will report back if I get anything, but I think we *could* get away with putting in a `.nupkg` with just the `.nuspec` inside. It needs testing though.

corngood commented:
> It actually [supports multiple](https://github.com/NuGet/Home/wiki/%5BSpec%5D-Fallback-package-folders#environment-variable), but they need to be separated by semicolons:

Ah yeah, I don't know why I thought that.  I swear I looked it up in the code before, but it's definitely splitting on `;`.

> How so? I thought it only searched through stuff that dotnet restore pulled, so if there's any dependencies on csprojs or nix-built stuff then there wouldn't be any .nupkgs there for it to find.

It searches `$NUGET_PACKAGES`, which includes all the things that have been symlinked from dependencies (sdk packages, etc).  It looks at `.nupkg.metadata`, which is empty in packages from the nix store.  Interestingly, (unlike `dotnet restore) `dotnet tool restore` doesn't seem to fill in `.nupkg.metadata` with the source it used to download.

> The fake HTTP daemon wouldn't work because it'd still try to download .nupkgs from the daemon so it'd be the same as using a directory as a source.

I wasn't sure if it would actually download them or just enumerate the available versions.

> I think we could get away with putting in a .nupkg with just the .nuspec inside. It needs testing though.

This is a good idea to try.

corngood commented:
> I think we could get away with putting in a .nupkg with just the .nuspec inside. It needs testing though.

It does seem to work with an uncompressed zip containing only the nuspec.  Something like that would be safe to put in outputs.

corngood commented:
Result of `nixpkgs-review pr 327651` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages failed to build:</summary>
  <ul>
    <li>dotnetCorePackages.dotnet_9.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_9.runtime</li>
    <li>dotnetCorePackages.dotnet_9.sdk</li>
    <li>dotnetCorePackages.dotnet_9.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_9.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_9.vmr</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>135 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>alttpr-opentracker</li>
    <li>am2rlauncher</li>
    <li>avalonia-ilspy</li>
    <li>azure-functions-core-tools</li>
    <li>beatsabermodmanager</li>
    <li>bicep</li>
    <li>bililiverecorder</li>
    <li>blendfarm</li>
    <li>boogie</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>cavalier</li>
    <li>celeste64</li>
    <li>certdump</li>
    <li>csharp-ls</li>
    <li>csharpier</li>
    <li>csharprepl</li>
    <li>cyclonedx-cli</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>dotnet-aspnetcore</li>
    <li>dotnet-aspnetcore_7</li>
    <li>dotnet-aspnetcore_8</li>
    <li>dotnet-outdated</li>
    <li>dotnet-runtime</li>
    <li>dotnet-runtime_7</li>
    <li>dotnet-runtime_8</li>
    <li>dotnet-sdk</li>
    <li>dotnet-sdk_7</li>
    <li>dotnet-sdk_8</li>
    <li>dotnetCorePackages.aspnetcore_9_0</li>
    <li>dotnetCorePackages.dotnet_8.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_8.runtime</li>
    <li>dotnetCorePackages.dotnet_8.sdk</li>
    <li>dotnetCorePackages.dotnet_8.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_8.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_8.vmr</li>
    <li>dotnetCorePackages.nugetPackageHook</li>
    <li>dotnetCorePackages.patchNupkgs</li>
    <li>dotnetCorePackages.runtime_9_0</li>
    <li>dotnetCorePackages.sdk_6_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_3xx</li>
    <li>dotnetCorePackages.sdk_8_0_1xx</li>
    <li>dotnetCorePackages.sdk_9_0</li>
    <li>eventstore</li>
    <li>fable</li>
    <li>famistudio</li>
    <li>fantomas</li>
    <li>formula</li>
    <li>fsautocomplete</li>
    <li>galaxy-buds-client</li>
    <li>garnet</li>
    <li>gh-gei</li>
    <li>git-credential-manager</li>
    <li>github-runner</li>
    <li>gitversion</li>
    <li>godot3-mono</li>
    <li>godot3-mono-debug-server</li>
    <li>godot3-mono-export-templates</li>
    <li>godot3-mono-headless</li>
    <li>godot3-mono-headless.dev</li>
    <li>godot3-mono-headless.man</li>
    <li>godot3-mono-server</li>
    <li>godot3-mono.dev</li>
    <li>godot3-mono.man</li>
    <li>ilspycmd</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>jellyfin</li>
    <li>jetbrains.clion</li>
    <li>jetbrains.rider</li>
    <li>kavita</li>
    <li>knossosnet</li>
    <li>kryptor</li>
    <li>libation</li>
    <li>lidarr</li>
    <li>lubelogger</li>
    <li>lumafly</li>
    <li>marksman</li>
    <li>mqttmultimeter</li>
    <li>msbuild</li>
    <li>naps2</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>networkminer</li>
    <li>nexusmods-app</li>
    <li>nexusmods-app-unfree</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>openra</li>
    <li>opentabletdriver</li>
    <li>openutau</li>
    <li>osu-lazer</li>
    <li>pablodraw</li>
    <li>parabolic</li>
    <li>pbm</li>
    <li>pinta</li>
    <li>pre-commit</li>
    <li>pre-commit.dist</li>
    <li>prowlarr</li>
    <li>ps3-disc-dumper</li>
    <li>pupdate</li>
    <li>python311Packages.asteroid-filterbanks</li>
    <li>python311Packages.asteroid-filterbanks.dist</li>
    <li>python311Packages.clr-loader</li>
    <li>python311Packages.clr-loader.dist</li>
    <li>python311Packages.pythonnet</li>
    <li>python311Packages.pythonnet.dist</li>
    <li>python312Packages.asteroid-filterbanks</li>
    <li>python312Packages.asteroid-filterbanks.dist</li>
    <li>python312Packages.clr-loader</li>
    <li>python312Packages.clr-loader.dist</li>
    <li>python312Packages.pythonnet</li>
    <li>python312Packages.pythonnet.dist</li>
    <li>radarr</li>
    <li>readarr</li>
    <li>retrospy</li>
    <li>roslyn</li>
    <li>roslyn-ls</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>slskd</li>
    <li>sonarr</li>
    <li>space-station-14-launcher</li>
    <li>tagger</li>
    <li>technitium-dns-server</li>
    <li>tone</li>
    <li>torrentstream</li>
    <li>vintagestory</li>
    <li>wasabibackend</li>
    <li>xivlauncher</li>
  </ul>
</details>

corngood commented:
I've updated the description to match the latest version.  Now we:

- point `$NUGET_FALLBACK_PACKAGES` at the dependencies

We also unconditionally:

- create a 'stub' source and add it to `nuget.config`

In theory this would be optional if all dependencies were locked, but that's rare enough that I'm not sure it makes sense for it to be configurable yet.

If `linkNugetPackages' is set, we:

- link all `$NUGET_FALLBACK_PACKAGES` into `$NUGET_PACKAGES`

This is required to work around upstream bugs like https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator/issues/94

If `installable` is passed to `mkNugetDeps`, we:

-  make full .nupkg files in the source instead of stubs.

This is required for `dotnet tool install`, because it can only install from the .nupkg, and not from an installed package.  It's also required for some VMR dependencies.

corngood commented:
Result of `nixpkgs-review pr 327651` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
  </ul>
</details>
<details>
  <summary>141 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>alttpr-opentracker</li>
    <li>am2rlauncher</li>
    <li>avalonia-ilspy</li>
    <li>azure-functions-core-tools</li>
    <li>beatsabermodmanager</li>
    <li>bicep</li>
    <li>bililiverecorder</li>
    <li>blendfarm</li>
    <li>boogie</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>cavalier</li>
    <li>celeste64</li>
    <li>certdump</li>
    <li>csharp-ls</li>
    <li>csharpier</li>
    <li>csharprepl</li>
    <li>cyclonedx-cli</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>dotnet-aspnetcore</li>
    <li>dotnet-aspnetcore_7</li>
    <li>dotnet-aspnetcore_8</li>
    <li>dotnet-outdated</li>
    <li>dotnet-runtime</li>
    <li>dotnet-runtime_7</li>
    <li>dotnet-runtime_8</li>
    <li>dotnet-sdk</li>
    <li>dotnet-sdk_7</li>
    <li>dotnet-sdk_8</li>
    <li>dotnetCorePackages.aspnetcore_9_0</li>
    <li>dotnetCorePackages.dotnet_8.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_8.runtime</li>
    <li>dotnetCorePackages.dotnet_8.sdk</li>
    <li>dotnetCorePackages.dotnet_8.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_8.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_8.vmr</li>
    <li>dotnetCorePackages.dotnet_9.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_9.runtime</li>
    <li>dotnetCorePackages.dotnet_9.sdk</li>
    <li>dotnetCorePackages.dotnet_9.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_9.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_9.vmr</li>
    <li>dotnetCorePackages.nugetPackageHook</li>
    <li>dotnetCorePackages.patchNupkgs</li>
    <li>dotnetCorePackages.runtime_9_0</li>
    <li>dotnetCorePackages.sdk_6_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_3xx</li>
    <li>dotnetCorePackages.sdk_8_0_1xx</li>
    <li>dotnetCorePackages.sdk_9_0</li>
    <li>eventstore</li>
    <li>fable</li>
    <li>famistudio</li>
    <li>fantomas</li>
    <li>formula</li>
    <li>fsautocomplete</li>
    <li>galaxy-buds-client</li>
    <li>garnet</li>
    <li>gh-gei</li>
    <li>git-credential-manager</li>
    <li>github-runner</li>
    <li>gitversion</li>
    <li>godot3-mono</li>
    <li>godot3-mono-debug-server</li>
    <li>godot3-mono-export-templates</li>
    <li>godot3-mono-headless</li>
    <li>godot3-mono-headless.dev</li>
    <li>godot3-mono-headless.man</li>
    <li>godot3-mono-server</li>
    <li>godot3-mono.dev</li>
    <li>godot3-mono.man</li>
    <li>ilspycmd</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>jellyfin</li>
    <li>jetbrains.clion</li>
    <li>jetbrains.rider</li>
    <li>kavita</li>
    <li>knossosnet</li>
    <li>kryptor</li>
    <li>libation</li>
    <li>lidarr</li>
    <li>lubelogger</li>
    <li>lumafly</li>
    <li>marksman</li>
    <li>mqttmultimeter</li>
    <li>msbuild</li>
    <li>naps2</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>networkminer</li>
    <li>nexusmods-app</li>
    <li>nexusmods-app-unfree</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>openra</li>
    <li>opentabletdriver</li>
    <li>openutau</li>
    <li>osu-lazer</li>
    <li>pablodraw</li>
    <li>parabolic</li>
    <li>pbm</li>
    <li>pinta</li>
    <li>pre-commit</li>
    <li>pre-commit.dist</li>
    <li>prowlarr</li>
    <li>ps3-disc-dumper</li>
    <li>pupdate</li>
    <li>python311Packages.asteroid-filterbanks</li>
    <li>python311Packages.asteroid-filterbanks.dist</li>
    <li>python311Packages.clr-loader</li>
    <li>python311Packages.clr-loader.dist</li>
    <li>python311Packages.pythonnet</li>
    <li>python311Packages.pythonnet.dist</li>
    <li>python312Packages.asteroid-filterbanks</li>
    <li>python312Packages.asteroid-filterbanks.dist</li>
    <li>python312Packages.clr-loader</li>
    <li>python312Packages.clr-loader.dist</li>
    <li>python312Packages.pythonnet</li>
    <li>python312Packages.pythonnet.dist</li>
    <li>radarr</li>
    <li>readarr</li>
    <li>retrospy</li>
    <li>roslyn</li>
    <li>roslyn-ls</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>slskd</li>
    <li>sonarr</li>
    <li>space-station-14-launcher</li>
    <li>tagger</li>
    <li>technitium-dns-server</li>
    <li>tone</li>
    <li>torrentstream</li>
    <li>vintagestory</li>
    <li>wasabibackend</li>
    <li>xivlauncher</li>
  </ul>
</details>

Smaug123 commented:
(I'm currently on holiday and diligently not doing anything that looks like work, but would in principle be interested in 1. this PR and 2. being on the dotnet team!)

corngood commented:
@Smaug123 have a good holiday and feel free to make a PR to add yourself here: https://github.com/NixOS/nixpkgs/blob/a1159ff6c68901142ae41788387bc4c841f190b2/maintainers/team-list.nix#L211

Smaug123 commented:
By the way, don't take my "I'm interested" as a desire to block this PR on my account if it's ready to go in before I get back!

nixf-tidy-review commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -4,20 +4,17 @@
+{ name ? "${_args.pname}-${_args.version}"
 , pname ? name
 , enableParallelBuilding ? true
 , doCheck ? false
```
definition `doCheck` is not used

nixf-tidy-review commented on pkgs/build-support/dotnet/build-dotnet-module/default.nix:
```diff
@@ -4,20 +4,17 @@
-{ name ? "${args.pname}-${args.version}"
+{ name ? "${_args.pname}-${_args.version}"
 , pname ? name
 , enableParallelBuilding ? true
```
definition `enableParallelBuilding` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/combine-packages.nix:
```diff
@@ -1,5 +1,5 @@
 dotnetPackages:
-{ buildEnv, makeWrapper, lib }:
+{ buildEnv, makeWrapper, lib, symlinkJoin }:
```
definition `makeWrapper` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/common.nix:
```diff
@@ -63,30 +65,29 @@
-            dotnet new ${template} -n test -o .
-          '' + build);
+            buildInputs =
+              [ sdk ]
```
redundant parentheses

nixf-tidy-review commented on pkgs/development/compilers/dotnet/common.nix:
```diff
@@ -14,6 +14,9 @@
 , swiftPackages
 , darwin
 , icu
```
definition `icu` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -5,45 +5,61 @@ dotnetCombined = with dotnetCorePackages; combinePackages [ sdk_6_0 aspnetcore_7
+{ lib
+, config
+, recurseIntoAttrs
+, makeSetupHook
```
definition `makeSetupHook` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/packages.nix:
```diff
@@ -1,6 +1,9 @@
 { stdenv
```
definition `stdenv` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/stage0.nix:
```diff
@@ -21,21 +21,24 @@ let
-
-  patchNupkgs = pkgsBuildHost.callPackage ./patch-nupkgs.nix {};
+  dotnetSdk = callPackage bootstrapSdk {};
 
```
definition `deps` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/stage1.nix:
```diff
@@ -1,7 +1,6 @@
 { stdenv
 , lib
```
definition `lib` is not used

nixf-tidy-review commented on pkgs/development/compilers/dotnet/stage1.nix:
```diff
@@ -1,7 +1,6 @@
 { stdenv
```
definition `stdenv` is not used

nixf-tidy-review commented on pkgs/development/tools/godot/3/mono/default.nix:
```diff
@@ -1,7 +1,6 @@
-, mkNugetSource
 , mono
 , dotnet-sdk
 , writeText
```
definition `writeText` is not used

nixf-tidy-review reviewed (commented):
⚠️  Hi! [nixf-tidy](https://github.com/nix-community/nixd/tree/main/libnixf) code linter found some issues in your touched files. If you believe I'm reporting false positives, feel free to ping <code>@inclyc</code>. 

I'm a review bot that based on git-diff thus there might be comments to previous code. Our nixpkgs CI will skip them, but you can also fix them if you like!

GGG-KILLER reviewed (approved):
LGTM! Sorry for the delay!

corngood commented:
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>147 packages built:</summary>
  <ul>
    <li>ArchiSteamFarm</li>
    <li>alttpr-opentracker</li>
    <li>am2rlauncher</li>
    <li>avalonia-ilspy</li>
    <li>azure-functions-core-tools</li>
    <li>beatsabermodmanager</li>
    <li>bicep</li>
    <li>bililiverecorder</li>
    <li>blendfarm</li>
    <li>boogie</li>
    <li>btcpayserver</li>
    <li>btcpayserver-altcoins</li>
    <li>cavalier</li>
    <li>celeste64</li>
    <li>certdump</li>
    <li>csharp-ls</li>
    <li>csharpier</li>
    <li>csharprepl</li>
    <li>cyclonedx-cli</li>
    <li>dafny</li>
    <li>denaro</li>
    <li>depotdownloader</li>
    <li>discordchatexporter-cli</li>
    <li>dotnet-aspnetcore</li>
    <li>dotnet-aspnetcore_7</li>
    <li>dotnet-aspnetcore_8</li>
    <li>dotnet-outdated</li>
    <li>dotnet-runtime</li>
    <li>dotnet-runtime_7</li>
    <li>dotnet-runtime_8</li>
    <li>dotnet-sdk</li>
    <li>dotnet-sdk_7</li>
    <li>dotnet-sdk_8</li>
    <li>dotnetCorePackages.aspnetcore_9_0</li>
    <li>dotnetCorePackages.dotnet_8.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_8.runtime</li>
    <li>dotnetCorePackages.dotnet_8.sdk</li>
    <li>dotnetCorePackages.dotnet_8.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_8.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_8.vmr</li>
    <li>dotnetCorePackages.dotnet_9.aspnetcore</li>
    <li>dotnetCorePackages.dotnet_9.runtime</li>
    <li>dotnetCorePackages.dotnet_9.sdk</li>
    <li>dotnetCorePackages.dotnet_9.sdk.artifacts</li>
    <li>dotnetCorePackages.dotnet_9.sdk.packages</li>
    <li>dotnetCorePackages.dotnet_9.vmr</li>
    <li>dotnetCorePackages.nugetPackageHook</li>
    <li>dotnetCorePackages.patchNupkgs</li>
    <li>dotnetCorePackages.runtime_9_0</li>
    <li>dotnetCorePackages.sdk_6_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_1xx</li>
    <li>dotnetCorePackages.sdk_7_0_3xx</li>
    <li>dotnetCorePackages.sdk_8_0_1xx</li>
    <li>dotnetCorePackages.sdk_9_0</li>
    <li>eventstore</li>
    <li>fable</li>
    <li>famistudio</li>
    <li>fantomas</li>
    <li>formula</li>
    <li>fsautocomplete</li>
    <li>galaxy-buds-client</li>
    <li>garnet</li>
    <li>gh-gei</li>
    <li>git-credential-manager</li>
    <li>github-runner</li>
    <li>gitversion</li>
    <li>godot3-mono</li>
    <li>godot3-mono-debug-server</li>
    <li>godot3-mono-export-templates</li>
    <li>godot3-mono-headless</li>
    <li>godot3-mono-headless.dev</li>
    <li>godot3-mono-headless.man</li>
    <li>godot3-mono-server</li>
    <li>godot3-mono.dev</li>
    <li>godot3-mono.man</li>
    <li>ilspycmd</li>
    <li>inklecate</li>
    <li>jackett</li>
    <li>jellyfin</li>
    <li>jetbrains.clion</li>
    <li>jetbrains.rider</li>
    <li>kavita</li>
    <li>knossosnet</li>
    <li>kryptor</li>
    <li>libation</li>
    <li>lidarr</li>
    <li>lubelogger</li>
    <li>lumafly</li>
    <li>marksman</li>
    <li>mqttmultimeter</li>
    <li>msbuild</li>
    <li>naps2</li>
    <li>nbxplorer</li>
    <li>netcoredbg</li>
    <li>networkminer</li>
    <li>nexusmods-app</li>
    <li>nexusmods-app-unfree</li>
    <li>nuget-to-nix</li>
    <li>omnisharp-roslyn</li>
    <li>openra</li>
    <li>opentabletdriver</li>
    <li>openutau</li>
    <li>osu-lazer</li>
    <li>pablodraw</li>
    <li>parabolic</li>
    <li>pbm</li>
    <li>pinta</li>
    <li>pre-commit</li>
    <li>pre-commit.dist</li>
    <li>prowlarr</li>
    <li>ps3-disc-dumper</li>
    <li>pupdate</li>
    <li>python311Packages.asteroid-filterbanks</li>
    <li>python311Packages.asteroid-filterbanks.dist</li>
    <li>python311Packages.clr-loader</li>
    <li>python311Packages.clr-loader.dist</li>
    <li>python311Packages.pyannote-audio</li>
    <li>python311Packages.pyannote-audio.dist</li>
    <li>python311Packages.pythonnet</li>
    <li>python311Packages.pythonnet.dist</li>
    <li>python312Packages.asteroid-filterbanks</li>
    <li>python312Packages.asteroid-filterbanks.dist</li>
    <li>python312Packages.clr-loader</li>
    <li>python312Packages.clr-loader.dist</li>
    <li>python312Packages.pyannote-audio</li>
    <li>python312Packages.pyannote-audio.dist</li>
    <li>python312Packages.pythonnet</li>
    <li>python312Packages.pythonnet.dist</li>
    <li>radarr</li>
    <li>readarr</li>
    <li>retrospy</li>
    <li>roslyn</li>
    <li>roslyn-ls</li>
    <li>ryujinx</li>
    <li>scarab</li>
    <li>slskd</li>
    <li>sonarr</li>
    <li>space-station-14-launcher</li>
    <li>tagger</li>
    <li>technitium-dns-server</li>
    <li>tone</li>
    <li>torrentstream</li>
    <li>vintagestory</li>
    <li>wasabibackend</li>
    <li>whisper-ctranslate2</li>
    <li>whisper-ctranslate2.dist</li>
    <li>xivlauncher</li>
  </ul>
</details>

corngood commented:
This is actually breaking some tests, like `nexusmods-app.tests.help`:

```
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Extensions.Hosting.Abstractions, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.Extensions.Hosting.Abstractions, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
/build/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 1:     5 Aborted                 (core dumped) /nix/store/riv0cmlbvkfydvglbgcc5iip6fdhjnm7-nexusmods-app-0.4.1/bin/NexusMods.App --help
```

Working on that now.

corngood committed:
nexusmods-app: publish only NexusMods.App.csproj

corngood commented on pkgs/by-name/ne/nexusmods-app/package.nix:
```diff
@@ -31,7 +31,13 @@ buildDotnetModule rec {
+  # example: Microsoft.Extensions.Hosting.dll 6.0.0 instead of 8.0.0.
+  # https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/7.0/solution-level-output-no-longer-valid
+  # TODO: do something about this in buildDotnetModule
+  projectFile = "src/NexusMods.App/NexusMods.App.csproj";
```
@l0b0 @MattSturgeon this reduces the output size significantly (~600->100MiB).  The tests still pass, but I'm wondering if there's anything we should be concerned about losing here.

MattSturgeon replied:
Thanks for the ping! We found in #318647 that we needed to use the csproj to avoid other issues too.

I'm not very knowledgeable on dotnet, but I think the solutions are generated _from_ the project files anyway; so this should be fine.

corngood commented:
I fixed another problem with `nexusmods-app`, and now all tests in the packages built by `nixpkgs-review` pass on linux.  I also updated the docs to reflect the `projectReferences` change.

corngood commented:
I'm planning on merging this soon.  I just wanted to check out some closure sizes.  Some things will be larger uncompressed, but should be similar compressed.

I also checked a few packages to confirm that build times are improved, and it seems likely to save in the neighborhood of 30s from restore in a package with a reasonable number of dependencies.
